jenkins/prepare-custom-leonardo-jupyter-dataproc-image.sh

```diff
@@ -0,0 +1,173 @@
+            log "Retry $i/$retries exited $exit, no more retries left."
+            break
+        fi
+        log "Retry $i/$retries exited $exit, retrying in $wait seconds..."
+        sleep $wait
+    done
+    return 1
+}
+
+function log() {
+    printf '[%s]: %s\n' \
+        "$(date +'%Y-%m-%dT%H:%M:%S%z')" \
+        "$*"
+}
+
+function betterAptGet() {
+    if ! { apt-get update 2>&1 || echo E: update failed; } | grep -q '^[WE]:'; then
+        return 0
+    else
+        return 1
+    fi
+}
+
+#
+# Main
+#
+log 'Installing prerequisites...'
+
+# Obtain the latest valid apt-key.gpg key file from https://packages.cloud.google.com to work
+# around intermittent apt authentication errors. See:
+# https://cloud.google.com/compute/docs/troubleshooting/known-issues
+retry 5 curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -
+retry 5 apt-key update
+
+# install Docker
+# https://docs.docker.com/install/linux/docker-ce/debian/
+export DOCKER_CE_VERSION="18.06.2~ce~3-0~debian"
+
+retry 5 betterAptGet
+retry 5 apt-get install -y -q \
+    apt-transport-https \
+    ca-certificates \
+    curl \
+    gnupg2 \
+    software-properties-common
+
+log 'Adding Docker package sources...'
+
+# shellcheck disable=SC1091
+os_dist_name="$(. /etc/os-release; echo "$ID")"
+os_dist_code_name="$(lsb_release -cs)"
+os_dist_release_channel="stable"
+os_dist_arch="amd64"
+
+docker_gpg_key_url="https://download.docker.com/linux/${os_dist_name:?}/gpg"
+docker_apt_repo_url="https://download.docker.com/linux/${os_dist_name:?}"
+
+retry 5 curl -fsSL "${docker_gpg_key_url:?}" | apt-key add -
+
+add-apt-repository \
+  "deb [arch=${os_dist_arch:?}] ${docker_apt_repo_url:?} \
+  ${os_dist_code_name:?} \
+  ${os_dist_release_channel:?}"
+
+log 'Installing Docker...'
+
+retry 5 betterAptGet
+retry 5 apt-get install -y -q docker-ce="${DOCKER_CE_VERSION:?}"
+
+log 'Installing Docker Compose...'
+
+# Install docker-compose
+# https://docs.docker.com/compose/install/#install-compose
+docker_compose_version_number="1.22.0"
+docker_compose_kernel_name="$(uname -s)"
+docker_compose_machine_hardware_name="$(uname -m)"
+docker_compose_binary_download_url="https://github.com/docker/compose/releases/download/${docker_compose_version_number:?}/docker-compose-${docker_compose_kernel_name:?}-${docker_compose_machine_hardware_name:?}"
+docker_compose_binary_download_target_filename="/usr/local/bin/docker-compose"
+
+retry 5 curl -L "${docker_compose_binary_download_url:?}" -o "${docker_compose_binary_download_target_filename:?}"
+chmod +x "${docker_compose_binary_download_target_filename:?}"
+
+# Pull docker image versions as of the time this script ran; this caches them in the
+# dataproc custom instance image.
+if [[ -n ${docker_image_var_names:?} ]]; then
+    for _docker_image_var_name in ${docker_image_var_names:?}
+    do
+        _docker_image="${!_docker_image_var_name:?}"
+        retry 5 docker pull "${_docker_image:?}"
+    done
+else
```

Comment:
does this else ever actually get run? I haven't seen the ? syntax before, but from playing around with it locally and from some documentation I read, it seems if any of the vars aren't defined it will exit without executing the else

Reply:
I think you're right -- `:?` probably shouldn't be there in this case. I think it was intended for the script to continue even if there aren't docker images set.